File: legacy/planner/src/main/java/com/risingwave/sql/tree/DropView.java
/*
 * Licensed to Crate.io GmbH ("Crate") under one or more contributor
 * license agreements.  See the NOTICE file distributed with this work for
 * additional information regarding copyright ownership.  Crate licenses
 * this file to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.  You may
 * obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * However, if you have executed another commercial license agreement
 * with Crate these terms will supersede the license and you may use the
 * software solely pursuant to the terms of the relevant commercial agreement.
 */

package com.risingwave.sql.tree;

/**
 * The drop view statement extended from CrateDB. Note that we modify the grammar from CrateDB, such
 * that we allow drop only one table once.
 */
public final class DropView extends Statement {

  private final QualifiedName name;
  private final boolean ifExists;
  private final boolean materialized;

  public DropView(QualifiedName name, boolean ifExists, boolean materialized) {
    this.name = name;
    this.ifExists = ifExists;
    this.materialized = materialized;
  }

  public QualifiedName name() {
    return name;
  }

  public boolean ifExists() {
    return ifExists;
  }

  public boolean isMaterialized() {
    return materialized;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }

    DropView dropView = (DropView) o;

    if (ifExists != dropView.ifExists) {
      return false;
    }
    return name.equals(dropView.name);
  }

  @Override
  public int hashCode() {
    int result = name.hashCode();
    result = 31 * result + (ifExists ? 1 : 0);
    return result;
  }

  @Override
  public String toString() {
    return "DropView{" + "name=" + name + ", ifExists=" + ifExists + '}';
  }

  @Override
  public <R, C> R accept(AstVisitor<R, C> visitor, C context) {
    return visitor.visitDropView(this, context);
  }
}

File: src/frontend/src/handler/drop_view.rs
// Copyright 2023 RisingWave Labs
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use pgwire::pg_response::{PgResponse, StatementType};
use risingwave_common::error::Result;
use risingwave_sqlparser::ast::ObjectName;

use super::RwPgResponse;
use crate::binder::Binder;
use crate::catalog::root_catalog::SchemaPath;
use crate::handler::HandlerArgs;

pub async fn handle_drop_view(
    handler_args: HandlerArgs,
    table_name: ObjectName,
    if_exists: bool,
) -> Result<RwPgResponse> {
    let session = handler_args.session;
    let db_name = session.database();
    let (schema_name, table_name) = Binder::resolve_schema_qualified_name(db_name, table_name)?;
    let search_path = session.config().get_search_path();
    let user_name = &session.auth_context().user_name;

    let schema_path = SchemaPath::new(schema_name.as_deref(), &search_path, user_name);

    let view_id = {
        let reader = session.env().catalog_reader().read_guard();
        let (view, schema_name) =
            match reader.get_view_by_name(session.database(), schema_path, &table_name) {
                Ok((t, s)) => (t, s),
                Err(e) => {
                    return if if_exists {
                        Ok(RwPgResponse::empty_result_with_notice(
                            StatementType::DROP_MATERIALIZED_VIEW,
                            format!("view \"{}\" does not exist, skipping", table_name),
                        ))
                    } else {
                        Err(e.into())
                    }
                }
            };

        session.check_privilege_for_drop_alter(schema_name, &**view)?;

        view.id
    };

    let catalog_writer = session.env().catalog_writer();
    catalog_writer.drop_view(view_id).await?;

    Ok(PgResponse::empty_result(StatementType::DROP_VIEW))
}
